plugin-server/src/main/ingestion-queues/anonymous-event-buffer-consumer.ts

```diff
@@ -0,0 +1,105 @@
+    graphileQueue,
+    statsd,
+}: {
+    kafka: Kafka
+    producer: KafkaProducerWrapper
+    graphileQueue: GraphileQueue
+    statsd?: StatsD
+}) => {
+    /*
+        Consumes from the anonymous event topic, and enqueues the events for
+        processing at a later date as per the message header `processEventAt`.
+
+        We do this delayed processing to allow for the anonymous users that
+        these events are associated to be merged or identified as other
+        "identified" users, at which point to can denormalize data to improve
+        query performance.
+
+        On failure to enqueue we will fail the batch. only on successfully
+        enqueuing an entire batch to Graphile Worker (a PostgreSQL backed async
+        worker library).
+
+        TODO:
+
+        1. distinguish between operational errors and programming errors
+        2. ensure offset only updated if message pushed to DLQ
+    */
+
+    const consumer = kafka.consumer({ groupId: 'ingester' })
+    setupEventHandlers(consumer)
+
+    status.info('🔁', 'Starting anonymous event buffer consumer')
+
+    const eachBatch: EachBatchHandler = async ({ batch, resolveOffset, heartbeat }) => {
+        status.info('🔁', 'Processing batch', { size: batch.messages.length })
+        for (const message of batch.messages) {
+            if (!message.value || !message.headers?.processEventAt || !message.headers?.eventId) {
+                status.warn('⚠️', `Invalid message for partition ${batch.partition} offset ${message.offset}.`, {
+                    value: message.value,
+                    processEventAt: message.headers?.processEventAt,
+                    eventId: message.headers?.eventId,
+                })
+                producer.queueMessage({ topic: KAFKA_EVENTS_DEAD_LETTER_QUEUE, messages: [message] })
+                resolveOffset(message.offset)
+                continue
+            }
+
+            let eventPayload: PluginEvent
+
+            try {
+                eventPayload = JSON.parse(message.value.toString())
+            } catch (error) {
+                status.warn('⚠️', `Invalid message for partition ${batch.partition} offset ${message.offset}.`, {
+                    error,
+                })
+                producer.queueMessage({ topic: KAFKA_EVENTS_DEAD_LETTER_QUEUE, messages: [message] })
+                resolveOffset(message.offset)
+                continue
+            }
+
+            const job = {
+                eventPayload: eventPayload,
+                timestamp: Number.parseInt(message.headers.processEventAt.toString()),
+                jobKey: message.headers.eventId.toString(), // Ensure we don't create duplicates
+            }
+
+            status.debug('⬆️', 'Enqueuing anonymous event for processing', { job })
+
+            await graphileQueue.enqueue(JobName.BUFFER_JOB, job)
+
+            // Resolve the offset such that, in case of errors further in
+            // the batch, we will not process these again
```

Comment:
I think this comment is a bit misguided?

From kafkajs docs:

> "In case of errors, the consumer will automatically commit the resolved offsets."

Reply:
Oh is that not what this means?